File: api/generate.py
from http.server import BaseHTTPRequestHandler
import json, os, urllib.request

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()
    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = json.loads(self.rfile.read(length))
            script = body.get("script", "")
            platform = body.get("platform", "Instagram")
            niche = body.get("niche", "")
            goal = body.get("goal", "")
            api_key = os.environ.get("GEMINI_API_KEY", "")
            if not api_key:
                raise Exception("GEMINI_API_KEY not set")
            prompt = f"You are a hashtag expert. Analyze this video script and generate hashtags for {platform}. Niche: {niche}. Goal: {goal}. Script: {script}. Return ONLY valid JSON with these keys: branded, high_volume, mid_tier, niche, strategy_notes"
            data = json.dumps({"contents": [{"parts": [{"text": prompt}]}]}).encode()
            req = urllib.request.Request(f"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={api_key}", data=data, headers={"Content-Type": "application/json"}, method="POST")
            with urllib.request.urlopen(req) as res:
                result = json.loads(res.read())
                text = result["candidates"][0]["content"]["parts"][0]["text"]
                text = text.replace("```json", "").replace("```", "").strip()
                parsed = json.loads(text)
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(json.dumps(parsed).encode())
        except Exception as e:
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode())

File: api-js/generate.js
export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  if (req.method === 'OPTIONS') return res.status(200).end();
  
  const { script, platform, niche, goal } = req.body;
  const api_key = process.env.OPENROUTER_API_KEY;
  
  const prompt = `Generate hashtags for ${platform}. Niche: ${niche}. Goal: ${goal}. Script: ${script}. Return ONLY valid JSON with keys: branded, high_volume, mid_tier, niche, strategy_notes. Each except strategy_notes is a list of hashtags.`;
  
  const response = await fetch('https://openrouter.ai/api/v1/chat/completions', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${api_key}` },
    body: JSON.stringify({ model: 'meta-llama/llama-3.3-70b-instruct:free', messages: [{ role: 'user', content: prompt }] })
  });
  
  const data = await response.json();
  let text = data.choices[0].message.content;
  text = text.replace(/```json/g, '').replace(/```/g, '').trim();
  res.status(200).json(JSON.parse(text));
}